[HB-8351] Create java demo app

wnloopme opened the pull request.

wnloopme committed:
[HB-8351] Create java demo app

chauduyphanvu commented on java/.gitignore:
```diff
@@ -0,0 +1,239 @@
+/.idea/caches
```
I think this gitignore is too verbose for what it should normally be. I recommend using wildcard matching.

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,276 @@
+        onLogStateChangeListener.logState("Start initializing mediation");
+        ChartboostCore.initializeSdkFromJava(context, new SdkConfiguration(appId, new ArrayList<>(), new HashSet<>()), new ModuleObserver() {
+            @Override
+            public void onModuleInitializationCompleted(@NonNull ModuleInitializationResult moduleInitializationResult) {
+                if (moduleInitializationResult.getModuleId().equals(ChartboostMediationSdk.CORE_MODULE_ID)) {
+                    if (moduleInitializationResult.getException() == null) {
+                        onLogStateChangeListener.logState("Initialization finished");
+                        ChartboostMediationSdk.setTestMode(context, true);
+                        onInitializationCompleteListener.onInitializationCompleted();
+                    } else {
+                        onLogStateChangeListener.logState("Initialization exception %s");
+                        onLogStateChangeListener.logState(moduleInitializationResult.getException().getMessage());
+                    }
+                }
```
I think we need an equivalent of `onInitializationCompleteListener()` for the `else` case (or one that accepts an error object for differentiation); otherwise, the calling app would never hear back.

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,276 @@
+
+    public void loadBanner(String placementName, OnLogStateChangeListener onLogStateChangeListener, FrameLayout bannerLayout, Context context) {
+        onLogStateChangeListener.logState("Start loading banner.");
+        ChartboostMediationBannerAdView banner = new ChartboostMediationBannerAdView(
```
`final` here and wherever applicable

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,276 @@
+                createBannerAdListener(onLogStateChangeListener)
+        );
+
+        bannerLayout.addView(banner);
```
You might want to clear out the `bannerLayout` of any existing child Views before adding a new one.

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,276 @@
+            public void onAdLoaded(@NonNull ChartboostMediationFullscreenAdLoadResult chartboostMediationFullscreenAdLoadResult) {
+                onLogStateChangeListener.logState("Ad loaded");
+                fullscreenAds.put(placementName, chartboostMediationFullscreenAdLoadResult.getAd());
+                button.setEnabled(fullscreenAds.get(placementName) != null);
```
Architecturally, `AdController` shouldn't be interacting with the app UI by touching buttons or layouts directly. That logic should be abstracted to an intermediate layer. It's fine here since it's nothing big, but ideally I'd recommend compartmentalization.

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,276 @@
+
+    public void initializeChartboost(
+            String appId,
+            OnLogStateChangeListener onLogStateChangeListener,
```
State changes should be communicated via listeners (and not via logging). Having `OnLogStateChangeListener` means we're conflating state changes and log events. Ideally we should break it down into two separate things/flows.

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,276 @@
+            OnLogStateChangeListener onLogStateChangeListener
+    ) {
+        ChartboostMediationFullscreenAdQueue queue = ChartboostMediationFullscreenAdQueueManager.queue(context, placement);
+        queue.setQueueCapacity(46);// this will update queue capacity only to the maximum capacity received on initialization
```
Move this to a constant

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,276 @@
+
+public class AdController {
+
+    private final Map<String, ChartboostMediationFullscreenAd> fullscreenAds = new HashMap<>();
+    private final String TAG = "AdController";
+    private final Map<String, ChartboostMediationFullscreenAdQueue> adQueues = new HashMap<>();
```
We might want `ConcurrentHashMap`s. Nothing big since it's a demo but still, wild use cases exist out there.

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,276 @@
+            if (queue.hasNextAd()) {
+                onLogStateChangeListener.logState("Using queue to show the ad");
+                ad = queue.getNextAd();
+                if (ad != null) {
+                    // Be mindful that ads will need to be reattached to your ChartboostMediationFullscreenAdListener if you are listening to ad cycle events.
+                    ad.setListener(createFullscreenAdListener(placementName, correspondingShowButton, onLogStateChangeListener));
+                }
```
Handle the `null` case.

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,276 @@
+        ChartboostMediationFullscreenAd ad = fullscreenAds.get(placementName);
+        if (shouldUseQueue) {
+            ChartboostMediationFullscreenAdQueue queue = getAdQueue(placementName);
+            if (queue.hasNextAd()) {
```
Handle the case where the queue does NOT have a next ad.

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,276 @@
+            }
+        };
+    }
+}
```
Clear the queue(s).

thomasmso commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,276 @@
+                        ChartboostMediationSdk.setTestMode(context, true);
+                        onInitializationCompleteListener.onInitializationCompleted();
+                    } else {
+                        onLogStateChangeListener.logState("Initialization exception %s");
```
think the format specifier is expecting a value to be provided?

wnloopme committed:
[HB-8351] Refine after CR

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,295 @@
+
+            @Override
+            public void onFullScreenAdQueueUpdated(@NonNull ChartboostMediationFullscreenAdQueue adQueue, @NonNull AdLoadResult result, int numberOfAdsReady) {
+                uiLogsListener.add("Fullscreen ad queue has been updated for placement " + "placementName" + ". Number of ads ready to show: " + numberOfAdsReady);
```
This is using the literal string `"placementName"` instead of the actual `placementName`. There are other instances of this as well.

wnloopme replied:
thanks

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,295 @@
+    ) {
+        ChartboostMediationFullscreenAdQueue queue = getAdQueue(placementName);
+        if (shouldUseQueue && queue != null && !queue.isRunning()) {
+            queue.start();
```
The current logic starts/stops the queue every time `loadFullscreenAd` is called based on the `shouldUseQueue` flag. Starting repeatedly is okay, but stopping it here feels misplaced. The queue should ideally be started/stopped only when the user toggles the switch in the `MainActivity`. Loading an ad shouldn't implicitly stop the queue if the user intended it to be running.

wnloopme replied:
changed the logic

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,295 @@
+
+public class AdController {
+
+    private static final int UNCONFIGURED_AD_QUEUE_CAPACITY = 46;
```
Any reasoning for 46 specifically? I'd recommend adding a comment. Otherwise, keep this number small (3-5) for demo purposes. We aren't stress-testing here.

wnloopme replied:
With help of this big number and the comment 
![image](https://github.com/user-attachments/assets/37dbd26c-30f5-46c7-bb96-0c06c4016a0e)
I wanted to show, that it doesn't matter what is put here, because it will be trimmed by the max capacity shown received on init.
I agree to with suggestion.

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/MainActivity.java:
```diff
@@ -0,0 +1,111 @@
+        clearLogsButton.setOnClickListener(v -> clearLogs());
+    }
+
+    private void onLoadClick(String placementName, Button correspondingShowButton) {
+        adController.loadFullscreenAd(placementName, getBaseContext(), correspondingShowButton, this);
+    }
```
it looks like users can repeatedly click the "Load" button while an ad is already loading. This can lead to multiple load requests firing, which might be confusing or inefficient.

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/MainActivity.java:
```diff
@@ -0,0 +1,111 @@
+        Toast.makeText(getBaseContext(), "Initialization failed", Toast.LENGTH_LONG).show();
+    }
+
+    @Override
+    protected void onDestroy() {
+        super.onDestroy();
+        adController.clear();
+    }
```
Also destroy and null out your ad instances.

wnloopme replied:
is this not enough?
![image](https://github.com/user-attachments/assets/45cfbc86-5d50-411e-9c25-6f43bf32cc15)

chauduyphanvu replied:
That's only emptying the queues - the objects themselves continue to exist. We also want to nullify the ad instances.

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,295 @@
+                    fullscreenAds.put(placementName, chartboostMediationFullscreenAdLoadResult.getAd());
+                    button.setEnabled(fullscreenAds.get(placementName) != null);
+                } else {
+                    uiLogsListener.add(String.format("Error loading ad: %s", chartboostMediationFullscreenAdLoadResult.getError().getMessage()));
```
Here and elsewhere, I'd also get and log the error code. That allows for quicker lookups.

wnloopme committed:
[HB-8351] Refine after CR vol2

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,296 @@
+                if (moduleInitializationResult.getModuleId().equals(ChartboostMediationSdk.CORE_MODULE_ID)) {
+                    if (moduleInitializationResult.getException() == null) {
+                        uiLogsListener.add("Initialization finished");
+                        ChartboostMediationSdk.setTestMode(context, true);
```
I'd tie test mode to the app's debug `BuildConfig` rather than blankly setting it.

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,296 @@
+                            uiLogsListener.add("Banner loaded");
+                        } else {
+                            uiLogsListener.add(String.format("Error loading ad, code %s : ", chartboostMediationBannerAdLoadResult.getError().getCode(), chartboostMediationBannerAdLoadResult.getError().getMessage()));
+
```
You can `toString()` the error to get the full concatenated error string.

chauduyphanvu replied:
Also I'd include the placement name in log messages for easy associations.

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,296 @@
+
+public class AdController {
+
+    private static final int UNCONFIGURED_AD_QUEUE_CAPACITY = 5;
```
What does unconfigured mean? Like a default?

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/MainActivity.java:
```diff
@@ -0,0 +1,112 @@
+        adController.initializeChartboost(
+                getString(R.string.default_app_id),
+                this,
+                getBaseContext(),
```
I'd just use `this` here instead of the base Context since that's more conventional and safer.

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,296 @@
+public class AdController {
+
+    private static final int UNCONFIGURED_AD_QUEUE_CAPACITY = 5;
+    private final Map<String, ChartboostMediationFullscreenAd> fullscreenAds = new ConcurrentHashMap();
```
this is super overkill if you only ever use two placements, but it's fine to have it to showcase that you can do this (assuming other people will be using multiple placements).
however, you don't showcase how in an ad queue setting, you can just chain shows.
You can get into a situation where an interstitial loads and an extra one caches. on show finished, the next ad is already ready to show. you don't need to wait for the next one to cache.

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,296 @@
+    private static final int UNCONFIGURED_AD_QUEUE_CAPACITY = 5;
+    private final Map<String, ChartboostMediationFullscreenAd> fullscreenAds = new ConcurrentHashMap();
+    private final Map<String, ChartboostMediationFullscreenAdQueue> adQueues = new ConcurrentHashMap();
+    private boolean shouldUseQueue = false;
```
conflating the queue usage and the not queue usage makes interesting code swaps. Maybe it'll be better to have a "use queue" screen and a "not using queue" screen.
Since the APIs and usages are completely different.
It's also unclear what happens when you load an ad manually then start the queue. when you show, which ad do you actually show?

wnloopme replied:
I imitated the ChartboostMediationDemoCompose - there is just one screen with the queue toggle.

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/AdController.java:
```diff
@@ -0,0 +1,296 @@
+            @Override
+            public void onFullScreenAdQueueUpdated(@NonNull ChartboostMediationFullscreenAdQueue adQueue, @NonNull AdLoadResult result, int numberOfAdsReady) {
+                uiLogsListener.add("Fullscreen ad queue has been updated for placement " + placementName + ". Number of ads ready to show: " + numberOfAdsReady);
+                showButton.setEnabled(numberOfAdsReady > 0);
```
like here: if you stop the queue, number of ads ready goes to 0 but if you're in the middle of a non-queue'd load and it finishes, this could cause this to be disabled.
Just having it all in once place is probably more confusing than it's worth.

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/MainActivity.java:
```diff
@@ -0,0 +1,112 @@
+
+    @Override
+    protected void onDestroy() {
+        super.onDestroy();
```
seconding this: all your loaded ads are not unloaded.
It's also not the greatest that you made the ad controller held by the activity. If the activity is destroyed, ads could still be loading in flight. also, the listener to the old ad queue is the old activity (or has access to the old activity).
It is possible to show ads loaded from another activity, so we should handle that (we don't need to showcase it; just make sure we're not leaking memory and doing general android best practices). It may not be the best idea to even unload the loaded ads, but if you don't have a good way to unload it, then that's fine; just destroy the loaded ad.
However, you also do have to handle the ad queue as that just keeps going. We should probably showcase to stop the ad queue when it's no longer used.

bichenwang commented on java/gradle/libs.versions.toml:
```diff
@@ -0,0 +1,22 @@
+[versions]
+agp = "8.5.1"
```
oh you have half the project with lib versions. why not convert the rest?

bichenwang commented on java/ChartboostMediationDemoJava/src/main/res/layout/log_item.xml:
```diff
@@ -0,0 +1,9 @@
+    android:layout_height="wrap_content"
+    android:padding="8dp"
+    android:textColor="@android:color/black"
+    tools:text="Some short text to be displayed" />
```
is this really what you want the default text to be?

wnloopme committed:
[HB-8351] Separate ManualLoadActivity and QueuedAdsActivity

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/JavaDemoApp.java:
```diff
@@ -0,0 +1,51 @@
+                        Toast.makeText(getBaseContext(), "Initialization failed", Toast.LENGTH_LONG).show();
+                    }
+                }
+
```
remove extra lines

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/Event.java:
```diff
@@ -0,0 +1,48 @@
+package com.example.chartboost.mediation.sdk.demo.java; // Or a suitable package
```
?

chauduyphanvu reviewed (commented):
general feedback: there are significant portions of code for creating ad listeners (`createFullscreenAdListener`, `createFullscreenAdShowListener`, `bannerAdListener`, `bannerAdLoadListener`) and UI logging (`addUiLog`, `_uiLogs` LiveData handling) that are nearly identical in both ViewModels.

bichenwang commented on java/ChartboostMediationDemoJava/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,36 @@
+        android:roundIcon="@mipmap/ic_launcher_round"
+        android:supportsRtl="true"
+        android:theme="@style/Theme.ChartboostMediationDemoJava"
+        tools:targetApi="31">
```
you're targeting 34 elsewhere?

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/manualload/ManualLoadViewModel.java:
```diff
@@ -0,0 +1,280 @@
+        }
+    }
+
+    private void loadAd(String placementName, Runnable onAdClosed, Consumer<ChartboostMediationFullscreenAd> onAdLoaded, Context context) {
```
In Java, it's a lot cleaner to name your callbacks and your anonymous methods. While it's fine for now, this makes it harder to maintain in the future.

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/manualload/ManualLoadViewModel.java:
```diff
@@ -0,0 +1,280 @@
+        );
+    }
+
+    public ChartboostMediationFullscreenAdShowListener getFullscreenAdShowListener(String placementName) {
```
So in kotlin, so you know how we use `val` and `var` right? Any time we use `val` in kotlin, in Java we generally want to use `final` in param names and variable declarations
So this would ideally be `public ChartboostMediationFullscreenAdShowListener getFullscreenAdShowListener(final String placementName) {` for example.

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/manualload/ManualLoadViewModel.java:
```diff
@@ -0,0 +1,280 @@
+
+    private ChartboostMediationFullscreenAdLoadListener createFullscreenAdLoadListener(
+            String placementName,
+            Consumer<ChartboostMediationFullscreenAd> onAdLoaded
```
The load result may be useful higher up, though I suppose we only print for now.
In theory, we could showcase the load id and/or other params from the load result.

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/manualload/ManualLoadViewModel.java:
```diff
@@ -0,0 +1,280 @@
+
+    private ChartboostMediationFullscreenAdListener createFullscreenAdListener(
+            String placementName,
+            Runnable onAdClosed
```
why is this a runnable? Some are lambdas and functions and some are Runnables? 
Would it make more sense to mostly use direct named callbacks? I mean it looks gross when you have to follow it all over the place, but that's java for you ...

wnloopme replied:
switched to use listeners and calling them back

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/queuedads/QueuedAdsViewModel.java:
```diff
@@ -0,0 +1,274 @@
+    private final MutableLiveData<List<String>> _uiLogs = new MutableLiveData<>(new ArrayList<>());
+    public final LiveData<List<String>> logs = _uiLogs;
+
+    private final Map<String, ChartboostMediationFullscreenAdQueue> queues = new HashMap<>();
```
you can certainly have this here, but the ad queue manager also keeps track of this exact map

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/queuedads/QueuedAdsViewModel.java:
```diff
@@ -0,0 +1,274 @@
+    private Boolean isSetup = false;
+    private String interstitialPlacement = "";
+    private String rewardedPlacement = "";
+    private static final int DEFAULT_QUEUE_CAPACITY = 5;
```
let's call out the fact that setting this outside the bounds of init will cap it to the amount set in init in a comment (or javadoc style comment) specifically for this variable.

wnloopme replied:
I've already put it at place of setting the queue 
        queue.setQueueCapacity(DEFAULT_QUEUE_CAPACITY);// this will update queue capacity only to the maximum capacity received on initialization

bichenwang replied:
Yes but if you put it as a javadoc, you can also see it in android studio when you hover over the variable.

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/queuedads/QueuedAdsViewModel.java:
```diff
@@ -0,0 +1,274 @@
+
+
+    public void setup(String interstitialPlacement, String rewardedPlacement) {
+        if (!isSetup) {
```
this is never false?

bichenwang reviewed (approved):
I like how kotlin-esque this turned out to be.
It would be super nice if we didn't have 90% the same code in the queued vs manual classes, but that can be an exercise for later. Or maybe Lucy could tackle that?

wnloopme committed:
[HB-8351] Extract BaseAdsViewModel, delete empty lines

wnloopme committed:
[HB-8351] Extract BaseAdsActivity

wnloopme committed:
[HB-8351] Addressing some warnings

wnloopme committed:
[HB-8351] Make functions parameters final

wnloopme committed:
[HB-8351] Transform to use callbacks

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/BaseAdsActivity.java:
```diff
@@ -0,0 +1,80 @@
+
+    protected void loadBanner(final FrameLayout bannerFrameLayout) {
+        viewModel.addToUiLogs("Loading banner started");
+        String bannerPlacementName = getString(R.string.banner_placement);
```
`final` is also used as `val`s in local / instance / class variables, too

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/BaseAdsActivity.java:
```diff
@@ -0,0 +1,80 @@
+
+        bannerFrameLayout.removeAllViews();
+        bannerFrameLayout.addView(banner);
+        banner.loadFromJava(
```
from an architecture perspective, you don't want to expose the load listener to the activity and you generally want to handle all of the logic in the view model itself, not the activity. That way, if in the future loading is changed, all activities will get said update.
Like you want to be like
`viewModel.loadBanner(banner)` so you don't need the load listener here. If you need an update from the view model, you can always observe a value from the view model.

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/BaseAdsViewModel.java:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    public void setup(final String interstitialPlacement, final String rewardedPlacement) {
+        if (!isSetup) {
```
I still don't see the value of `isSetup`
You're not blocking anything from this; you're not changing anything by having it.
It's one thing if you're blocking loads if the view model isn't initialized, but right now, you're not really checking. you can still ostensibly load a placement of ""
If you wanted unchanged placements on initialization, you could always make these constructor arguments (but view models with constructors are annoying).

wnloopme replied:
removed

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/BaseAdsViewModel.java:
```diff
@@ -0,0 +1,154 @@
+        };
+    }
+
+    public void setup(final String interstitialPlacement, final String rewardedPlacement) {
```
a good chunk of the class is already annotated with this, but please have the @NonNull and @Nullable annotations. Yay for writing kotlin-y java! It also helps with static analysis so you get the "this could be null and you're hitting it unchecked" compiler warnings!

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/BaseAdsViewModel.java:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    protected void addUiLogs(final String... logs) {
+        List<String> updatedLogs = new ArrayList<>(_uiLogs.getValue());
```
why not keep a backing variable instead of creating a new array every time?

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/BaseAdsViewModel.java:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    @Override
+    protected void onCleared() {
```
you don't need to define this if this is the only thing it's doing...

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/CustomFullscreenAdListener.java:
```diff
@@ -0,0 +1,18 @@
+
+    void onAdShowFailure(final String placementName, @Nullable final Boolean hasAdQueued);
+
+    void onAdShowSuccess(final String placementName, @Nullable Integer numberOfAdsReady);
```
Why is this an `Integer` and not an `int`?

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/CustomFullscreenAdListener.java:
```diff
@@ -0,0 +1,18 @@
+
+    void onAdClosed(final String placementName, final @NonNull ChartboostMediationFullscreenAd chartboostMediationFullscreenAd, final @Nullable ChartboostMediationAdException e, @Nullable Boolean hasNextAdQueued);
+
+    void onAdShowFailure(final String placementName, @Nullable final Boolean hasAdQueued);
```
why is this a `Boolean` and not a `boolean`?

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/JavaDemoApp.java:
```diff
@@ -0,0 +1,48 @@
+
+public class JavaDemoApp extends Application {
+
+    private static final String TAG = "JavaDemoApp";
```
A thing that has helped me in the past:
TAG = [class].simpleName or .name or something to that effect. (short name? forgot exactly off the top of my head). that way if the class name ever changes, it automatically changes with it. (you don't have to do this here; just as a note in the future).

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/MainActivity.java:
```diff
@@ -0,0 +1,38 @@
+            openManualLoadAds();
+        });
+        binding.queuedAdsBtn.setOnClickListener((button) -> {
+            openQueuedAdsClicked();
```
please be consistent with the naming of these two methods...

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/manualload/ManualLoadActivity.java:
```diff
@@ -0,0 +1,56 @@
+
+        setupLogsRecyclerView(binding.logsRv);
+
+        viewModel = new ViewModelProvider(this).get(ManualLoadViewModel.class);
```
could be easier depending on how many times you hard cast it to ManualLoadViewModel to keep the variable in this class and cast it once on get.
You're introducing the possibility of a class cast exception anywhere you do the cast.

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/manualload/ManualLoadViewModel.java:
```diff
@@ -0,0 +1,213 @@
+    private final MutableLiveData<ManualLoadUIStateModel> _uiState = new MutableLiveData(ManualLoadUIStateModel.initialState());
+
+    private final MutableLiveData<Event<ChartboostMediationFullscreenAd>> _showAdEvent = new MutableLiveData<>();
+    public final LiveData<Event<ChartboostMediationFullscreenAd>> showAdEvent = _showAdEvent;
```
pretty sure you're still supposed to have getters in java and not public fields. So in java, this makes it so you don't need the _ at all since the backing field is `showAdEvent` and you'll just have the getter for `getShowAdEvent` which returns a not mutable `LiveData<Event<ChartboostMediationFullscreenAd>>`

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/queuedads/CustomAdQueueListener.java:
```diff
@@ -0,0 +1,5 @@
+package com.example.chartboost.mediation.sdk.demo.java.queuedads;
+
+public interface CustomAdQueueListener {
```
what makes this `custom`? Please be more descriptive than "this is special because I said so."

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/BaseAdsViewModel.java:
```diff
@@ -0,0 +1,154 @@
+    private final MutableLiveData<List<String>> _uiLogs = new MutableLiveData<>(new ArrayList<>());
+    public final LiveData<List<String>> uiLogs = _uiLogs;
+
+    public final ChartboostMediationBannerAdViewListener bannerAdListener;
```
should these be public?

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/BaseAdsViewModel.java:
```diff
@@ -0,0 +1,154 @@
+public abstract class BaseAdsViewModel extends ViewModel {
+
+    protected Boolean isSetup = false;
+    protected String interstitialPlacement = "";
```
I'm surprised there isn't a banner placement?

bichenwang commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/queuedads/QueuedAdsUIState.java:
```diff
@@ -0,0 +1,56 @@
+package com.example.chartboost.mediation.sdk.demo.java.queuedads; // Adjust package if needed
```
is this a comment for us or for people looking at this?

wnloopme replied:
🤦 for me- for not noticing it before.

bichenwang reviewed (approved):
most of the comments are just things you should keep your eyes out for the next time. feel free to do however much of them you want. 
the named callbacks actually turned out better than expected! good job!

chauduyphanvu commented on java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/BaseAdsActivity.java:
```diff
@@ -0,0 +1,80 @@
+    abstract protected void setupObservers();
+
+    protected void updateLogs(final List<String> logs) {
+        int size = logs.size();
```
`final`

wnloopme committed:
[HB-8351] CR refinement:
 - rename CustomAdQueueListener.java and CustomFullscreenAdListener.java
 - move logic of creating banner to ViewModel
 - delete obsolete isSetup variable
 - expose getters for livedatas instead of having backing fields
